keps/sig-cluster-lifecycle/kubeadm/4656-add-kubelet-instance-configuration/README.md

```diff
@@ -0,0 +1,290 @@
+* No longer need to add the `kubeadm.alpha.kubernetes.io/cri-socket` annotation.
+* If the CRI socket provided in the kubeadm configuration is set, it is used first and the `/var/lib/kubelet/instance-config.yaml` configuration file is generated based on it, overwriting the kubelet configuration downloaded from the global configuration; If no CRI socket is specified, the socket is automatically detected on the node and `/var/lib/kubelet/instance-config.yaml` is generated based on it.
+
+kubeadm reset:
+
+* There is no need to do anything, according to the existing process, we get CRISocketPath before deleting the /var/lib/kubelet directory, and after deleting the `/var/lib/kubelet` directory, `/var/lib/kubelet/instance-config.yaml` will also be cleaned up.
+
+kubeadm upgrade:
+
+* `kubeadm upgrade node/apply` will check the `--container-runtime-endpoint` args in the `/var/lib/kubelet/kubeadm-flags.env` file and generate `/var/lib/kubelet/instance-config.yaml` based on them, and override the `ContainerRuntimeEndpoint` field to `/var/lib/kubelet/config.yaml`.
+
+**In the Beta phase, the feature gate is enabled by default. If feature gate is disabled, kubeadm subcommands will not be changed, when the feature gate is enabled, the kubeadm subcommands change as follows:** 
+
+* `kubeadm upgrade apply/node`  will use `/var/lib/kubelet/instance-config.yaml`, and override the `ContainerRuntimeEndpoint` field to `/var/lib/kubelet/config.yaml`.
+
+**In the Beta phase, the feature gate is enabled by default and cannot be disabled. the kubeadm subcommands change as follows:** 
+
+* `kubeadm upgrade apply/node` will use `/var/lib/kubelet/instance-config.yaml` and remove CRI annotations from Node object and remove `--container-runtime-endpoint` args in `/var/lib/kubelet/kubeadm-flags.env` file.
+
+### Test Plan
+
+[x] I/we understand the owners of the involved components may require updates to
+existing tests to make this code solid enough prior to committing the changes necessary
+to implement this enhancement.
+
+##### Prerequisite testing updates
+
+##### Unit tests
+
+At least the following kubeadm packages will require updates and new unit tests:
+
+* `cmd/kubeadm/app/cmd/phases/init`
+
+* `cmd/kubeadm/app/phases/kubelet`
+
+* `cmd/kubeadm/app/phases/upgrade`
+
+* `cmd/kubeadm/app/cmd/phases/join`
+
+##### Integration tests
+
+- N/A
+
+##### e2e tests
+
+* A new e2e test will be added by using the kinder tool.
+
+### Graduation Criteria
+
+#### Alpha
+
+- Use `NodeLocalCRISocket` feature gate to implement features.
+- Add corresponding e2e tests.
+- Added documentation for feature gates.
+
+#### Beta
+
+* Make feature gate to be enabled by default.
+
+- Gather feedback from developers and surveys.
+- Update the feature gate documentation.
+- Implement changes in kubeadm upgrade apply/node Beta phase.
+
+#### GA
+
+- Gather feedback from developers and surveys.
+- Implement changes in kubeadm upgrade apply/node GA phase.
+- Update the phases documentation.
```

Comment:
After GA, we may not add the old alpha annotation.
- We may deprecate the annotation in Beta?

We can keep the annotation if the node is upgraded from older version, and it can be removed manually if they want then.

Reply:
> We can keep the annotation if the node is upgraded from older version, and it can be removed manually if they want then.

Yes, this is what we want to do, once the feature gate is enabled, we will not set the CRI socket annotation during `kubeadm init/join`, nor remove it on upgrade.

So I think it has nothing to do with which phase we are in, because we already have feature gates to help us control version skew.

Reply:
please add entry in the above list to remove the annotation from here too:
https://kubernetes.io/docs/reference/labels-annotations-taints/#kubeadm-alpha-kubernetes-io-cri-socket
once the FG goes GA.

this page also needs an update:
https://kubernetes.io/docs/tasks/administer-cluster/migrating-from-dockershim/change-runtime-containerd/
to tell the user to update instance-config instead of the annotation.

Reply:
this was not updated.

Reply:
I appended them in subsequent lines.